Lors de ce TP, nous allons explorer la notion d'information mutuelle dans le cadre d'une tâche de classification de mails :  un mail est-il un spam ou un ham . 



# Les données

Récupérer les données (lingspam.tgz) et décompresser l'archive. 
- regarder le contenu de l'archive
- il existe différentes versions des données, lesquelles ? 

Voici du code pour lire et mettre en forme les donnéés (avec les imports aussi): 

In [36]:
import pickle, os 
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.tree import DecisionTreeClassifier, export_graphviz

train_texts, test_texts, train_labels, test_labels = pickle.load(open("lingspam.p", "rb" ))


- De quels types sont les variables  train_texts, test_texts, train_labels, test_labels ? 
- leurs dimensions ? 
- De quels types sont les données stockées ? 
- Combien y-a-t-il d'exemple pour l'apprentissage et le test  ?

## Matrice terme/document 

Nous allons maintenant transformer les données en utilisant sklearn. Pour cela il existe un outil: CountVectorizer
Cet objet permet de générer une matrice de compte. Nous allons utiliser la version binaire: chaque caractéristique est binaire, indépendante des autres caractéristique et représente la présence (ou l'absence) d'un mot dans un mail.  Pour cela nous procédons en 2 étapes: 


In [37]:
import numpy
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(binary=True)
x_train = cv.fit_transform(train_texts) # étape 1 
x_test = cv.transform(test_texts) # étape 2 


Regarder la documentation de sklearn sur CountVectorizer et répondre aux questions suivantes

## Questions sur les données

- Comment expliquer ces deux étapes ? 
- Combien y-a-t-il de document dans les données de train ? 
- Combien y-a-t-il de document dans les données de test ? 
- Quel est le vocabulaire ? 
- Afficher le vecteur associé au premier document d'apprentissage? 
- Quel est le type de l'objet x_train ? Expliquer ce choix ? 
- Comment trouver le lien entre un mot et son indice ? 

- cela permet de transformer les textes en matrice qui compte le nombre d'occurence de chaque token

In [38]:
print 'nb d element de train :', x_train.shape
print 'nb d element de test :', x_test.shape
voc = cv.vocabulary_

nb d element de train : (2604, 57752)
nb d element de test : (290, 57752)


In [22]:
print type(x_train)
print x_train[0]

<class 'scipy.sparse.csr.csr_matrix'>
  (0, 50070)	1
  (0, 41883)	1
  (0, 20948)	1
  (0, 16243)	1
  (0, 16311)	1
  (0, 1432)	1
  (0, 11805)	1
  (0, 22223)	1
  (0, 39743)	1
  (0, 41876)	1
  (0, 28323)	1
  (0, 14438)	1
  (0, 38781)	1
  (0, 7202)	1
  (0, 37211)	1
  (0, 29754)	1
  (0, 757)	1
  (0, 1011)	1
  (0, 1435)	1
  (0, 51425)	1
  (0, 8916)	1
  (0, 54082)	1
  (0, 25492)	1
  (0, 38579)	1
  (0, 29324)	1
  :	:
  (0, 22160)	1
  (0, 28107)	1
  (0, 5432)	1
  (0, 26395)	1
  (0, 42751)	1
  (0, 6201)	1
  (0, 8420)	1
  (0, 13637)	1
  (0, 49150)	1
  (0, 39291)	1
  (0, 17196)	1
  (0, 14333)	1
  (0, 22929)	1
  (0, 14730)	1
  (0, 18861)	1
  (0, 56129)	1
  (0, 53944)	1
  (0, 36493)	1
  (0, 16233)	1
  (0, 22337)	1
  (0, 11900)	1
  (0, 6830)	1
  (0, 44930)	1
  (0, 5073)	1
  (0, 3916)	1


# Arbre de décision

La classe pour apprendre un arbre de décision est DecisionTreeClassifier. Regarder la documentation de cette classe et commenter la ligne suivante: 

In [106]:
classifier = DecisionTreeClassifier(criterion="entropy", splitter='best', random_state = 0, max_depth=None)

- apprendre l'arbre de décision sur les données d'apprentissage
- comment connaitre la prédiction du modèle pour le premier document de test ? 
- la réponse du modèle est-elle correcte ? 
- Calculer le score de bonne prédiction sur les données de test. 
- Afficher les 10  caractéristiques sélectionnnés comme les plus importantes lors de la contruction de l'arbre de décision. 
- Regarder l'importance pour toutes les caractéristiques, que constatez-vous ? 


In [107]:
classifier.fit(x_train, train_labels)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=0, splitter='best')

In [108]:
classifier.score(x_test, test_labels)

0.95862068965517244

In [109]:
import operator
def top_k(k, imp):
    res = sorted(zip(range(len(imp)), imp), reverse=True, key=operator.itemgetter(1))
    return res[:k]

In [110]:
tk= top_k(50, classifier.feature_importances_)
print (classifier.feature_importances_!=0).sum()
for cp in tk:
    print cp[1], cv.get_feature_names()[cp[0]]
    


49
0.263709204449 language
0.177927297975 free
0.0777815507395 you
0.070937972508 university
0.0507699259446 linguistics
0.0292523492465 edu
0.0276051786827 english
0.0243224999614 your
0.0232897516966 remove
0.0208283972812 click
0.017417435449 research
0.0150030248667 syntax
0.0149088710052 share
0.0114560297806 http
0.0108131485559 de
0.00844281436105 pp
0.00750800744329 anyone
0.0074266466415 visit
0.00737689518175 fax
0.00712912615655 call
0.00697347556023 restoring
0.00690538745143 suggestions
0.00678823908332 interviews
0.00670094398132 ftp
0.00634679827869 via
0.00623496912229 gino
0.00619815334066 nl
0.00613300126313 role
0.00609663391423 exclusive
0.00591081751306 official
0.00528860354546 ll
0.00522811988669 price
0.0050500942987 equitybuilder
0.00460385995232 xxx
0.00452962389203 or
0.00450870404082 seems
0.00395774541051 springfield
0.00355940579297 how
0.00352743241026 sending
0.00344082061807 chosen
0.00326858751139 information
0.00231369357318 following
0.00214136253588

## Profondeur de l'arbre 

Effectuer un apprentissage, puis évaluer le modèle avec différentes profondeurs maximum d'arbre (1,2,5,10,15,20,100,None): 
- construire le classifieur
- calculer son score sur le test
Une profondeur maximum de *None*  signifie qu'il n'y a pas de limitation. 
- Que constatez vous ? 

In [82]:
clf = DecisionTreeClassifier(criterion="entropy", splitter='best', random_state=0, max_depth=None)

In [83]:
clf.fit(x_train, train_labels)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None, min_samples_leaf=1,
            min_samples_split=2, min_weight_fraction_leaf=0.0,
            presort=False, random_state=0, splitter='best')

In [84]:
clf.score(x_test, test_labels)

0.95862068965517244

* depth = 1 : 0.82758620689655171
* depth = 2 : 0.91724137931034477
* depth = 5 : 0.93103448275862066
* depth = 10 : 0.96206896551724141
* depth = 15 : 0.95862068965517244
* depth = 20 : 0.95862068965517244
* depth = 100 : 0.95862068965517244
* depth = None : 0.95862068965517244

On constate que le résultat ne s'améliore plus même qu'il réduit avant de stagner

## Visualisation de l'arbre
Il est possible de générer une figure qui représente l'arbre ainsi: 

In [123]:
from sklearn.tree import  export_graphviz
export_graphviz(clf, out_file = "unNomDeFichier", feature_names = cv.get_feature_names())


Puis vous pouvez convertir le fichier en pdf grâce à la commande unix suivante : 
 dot -Tpdf unNomDeFichier  -o maFigure.pdf

- Visualiser l'arbre 
- Regarder les seuils de décision  appliqués ? Expliquer les valeurs obtenues. 

# Information mutuelle 

A partir des données d'apprentissage, calculer: 
- l'entropie de la variable Y (qui représente la classe)
- l'information mutuelle entre Y et chaque caractéristique binaire
- quels sont les 10 caractéristiques (mots) ayant l'information mutuelle la plus élevée. 

Remarque: Pour accéder facilement aux comptes, il suffit d'apprendre un Bayésien Naif de type Bernoulli avec sklearn. L'objet résultant contient les comptes. Recouper ces résultats avec ceux obtenus avec l'arbre de décision. 
